src/Processors/Executors/ExecutorTasks.h

```diff
@@ -38,13 +38,18 @@ class ExecutorTasks
     /// Started thread count (allocated by `ConcurrencyControl`). Can increase during execution up to `num_threads`.
     size_t use_threads = 0;
 
+    /// Number of idle threads, changed with threads_queue.size().
+    std::atomic_size_t idle_threads = 0;
+
     /// This is the total number of waited async tasks which are not executed yet.
     /// sum(executor_contexts[i].async_tasks.size())
     size_t num_waiting_async_tasks = 0;
 
     /// A set of currently waiting threads.
     ThreadsQueue threads_queue;
 
+    const static size_t TOO_MANY_IDLE_THRESHOLD = 4;
```

Comment:
Please add a comment how you found constant = 4. Fine to document it as "Found by rolling dice.".